File: app/src/main/java/me/darshpratap/coaster/MainActivity.java
package me.darshpratap.coaster;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import me.darshpratap.coaster.api.PageInsightService;
import me.darshpratap.coaster.database.viewModel.CategoryViewModel;
import me.darshpratap.coaster.database.viewModel.ContentViewModel;
import me.darshpratap.coaster.database.viewModel.ResponseViewModel;
import me.darshpratap.coaster.models.db.Category;
import me.darshpratap.coaster.models.db.Response;
import retrofit2.Call;
import retrofit2.Callback;
import android.os.Bundle;
import android.util.Log;
import android.widget.LinearLayout;

import com.google.android.material.bottomsheet.BottomSheetBehavior;

public class MainActivity extends AppCompatActivity {
    private ResponseViewModel responseViewModel;
    private CategoryViewModel categoryViewModel;
    private ContentViewModel contentViewModel;
    private LinearLayout historyBottomSheet;
    private BottomSheetBehavior sheetBehavior;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        historyBottomSheet = findViewById(R.id.history_bottom_sheet);
        sheetBehavior = BottomSheetBehavior.from(historyBottomSheet);

        responseViewModel = ViewModelProviders.of(MainActivity.this).get(ResponseViewModel.class);
        categoryViewModel = ViewModelProviders.of(MainActivity.this).get(CategoryViewModel.class);
        contentViewModel = ViewModelProviders.of(MainActivity.this).get(ContentViewModel.class);

        PageInsightService.getInsightApi().getResponse().enqueue(new Callback<ResponseApi>() {
            @Override
            public void onResponse(Call<ResponseApi> call, retrofit2.Response<ResponseApi> response) {
                cascadedUpsert(response.body(), "shit: fix this soon");
            }

            @Override
            public void onFailure(Call<ResponseApi> call, Throwable t) {
                Log.d("shitshit", "Error: Request Failed " + t.getMessage());
            }
        });
    }

    private void cascadedUpsert(final ResponseApi responseApi, final String url) {
//        int id = responseViewModel.getId(url);
//        Log.d("shitid", String.valueOf(id));
//        // insert if id != 0 else update
//        if (id == 0) {
//            responseViewModel.insert(new Response(url));
//            id = responseViewModel.getId(url);
//            categoryViewModel.insert(new Category(responseApi.getPerformanceApi().getTitle(), responseApi.getPerformanceApi().getScore(), id));
//            categoryViewModel.insert(new Category(responseApi.getAccessibiltyApi().getTitle(), responseApi.getAccessibiltyApi().getScore(), id));
//            categoryViewModel.insert(new Category(responseApi.getBest_practicesApi().getTitle(), responseApi.getBest_practicesApi().getScore(), id));
//            categoryViewModel.insert(new Category(responseApi.getPwaApi().getTitle(), responseApi.getPwaApi().getScore(), id));
//            categoryViewModel.insert(new Category(responseApi.getSeoApi().getTitle(), responseApi.getSeoApi().getScore(), id));
//        } else {
//
//        }

        responseViewModel.getId(url).observe(this, new Observer<Integer>() {
            @Override
            public void onChanged(Integer id) {
                if (id == null) {
                    responseViewModel.insert(new Response(url));
                    categoryViewModel.insert(new Category(responseApi.getPerformanceApi().getTitle(), responseApi.getPerformanceApi().getScore(), id));
                }
            }
        });
    }
}

File: app/src/main/java/me/darshpratap/coaster/activity/MainActivity.kt
package me.darshpratap.coaster.activity

import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View.*
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.bottomsheet.BottomSheetBehavior
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.history_bottom_sheet.*
import me.darshpratap.coaster.R
import me.darshpratap.coaster.adapter.HistoryAdapter
import me.darshpratap.coaster.api.PageInsightService
import me.darshpratap.coaster.database.viewModel.CascadedOperations
import me.darshpratap.coaster.database.viewModel.HistoryWithCategoryViewModel
import me.darshpratap.coaster.models.api.ResponsePojo
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {
    private val TAG = "MainActivityTAG"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

//        TODO("check on text from input edit text")
//        val historyViewModel = ViewModelProviders.of(this).get(HistoryViewModel::class.java)
        val historyWithCategoryViewModel = ViewModelProviders.of(this).get(HistoryWithCategoryViewModel::class.java)
        val adapter = HistoryAdapter(this)
        val scale = resources.displayMetrics.density
        val bottomSheetBehavior = BottomSheetBehavior.from(history_bottom_sheet)

        // setting recycler view
        hbs_recycler_view.layoutManager = LinearLayoutManager(this)
        hbs_recycler_view.adapter = adapter

//        TODO("hide bottom sheet when scrolled")
//        hbs_recycler_view.addOnScrollListener(object: RecyclerView.OnScrollListener() {
//            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
//                if(dy != 0) {
//                    val scale = resources.displayMetrics.density
//                    BottomSheetBehavior.from(history_bottom_sheet).peekHeight = (50 * scale).toInt()
//                }
//            }
//        })
        bottomSheetBehavior.peekHeight = (50 * scale).toInt()
        bottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED

        // live data observer set-up
        historyWithCategoryViewModel.getHistoryWithCategory()?.observe(this, Observer {
            Log.d("shit HST", it.size.toString())
            adapter.submitList(it)
            hbs_recycler_view.smoothScrollToPosition(0)
        })

        insight_btn_desktop.setOnClickListener {
            val url = url_text_input.text.toString()
            if (!url.isEmpty()) {
                requestApi(url, "desktop")
            }
        }

        insight_btn_mobile.setOnClickListener {
            val url = url_text_input.text.toString()
            if (!url.isEmpty()) {
                requestApi(url, "mobile")
            }
        }
    }

    private fun requestApi(url: String, strategy: String) {
        disableButtons()
        PageInsightService.getInsightApi()?.getResponse(url, strategy)?.enqueue(object : Callback<ResponsePojo> {
            override fun onFailure(call: Call<ResponsePojo>, t: Throwable) {
                Toast.makeText(this@MainActivity, t.message, Toast.LENGTH_SHORT).apply {
                    setGravity(Gravity.TOP, 0, 60)
                    show()
                }
                enableButtons()
            }

            override fun onResponse(call: Call<ResponsePojo>, response: Response<ResponsePojo>) {
                enableButtons()
                if(response.body()?.performancePojo != null) {
                    val cascadedOperations = CascadedOperations()
                    cascadedOperations.cascadedInsert(application, url, strategy, response.body())
                } else {
                    Toast.makeText(this@MainActivity, "Invalid URL", Toast.LENGTH_SHORT).apply {
                        setGravity(Gravity.TOP, 0, 60)
                        show()
                    }
                }
//                Intent(this@MainActivity, InsightActivity::class.java).apply {
//                    startActivity(this)
//                }
            }

        })
    }

    private fun disableButtons() {
        loading.visibility = VISIBLE
        insight_icon.visibility = GONE
        insight_btn_desktop.isClickable = false
        insight_btn_desktop.backgroundTintList = ColorStateList.valueOf(Color.DKGRAY)
        insight_btn_mobile.backgroundTintList = ColorStateList.valueOf(Color.DKGRAY)
        insight_btn_mobile.isClickable = false
    }

    private fun enableButtons() {
        loading.visibility = GONE
        insight_icon.visibility = VISIBLE
        insight_btn_desktop.isClickable = true
        insight_btn_desktop.backgroundTintList = ColorStateList.valueOf(resources.getColor(R.color.colorAccent))
        insight_btn_mobile.backgroundTintList = ColorStateList.valueOf(resources.getColor(R.color.colorAccent))
        insight_btn_mobile.isClickable = true
    }
}